### ISSUE TITLE
OpenMP backend uses omp critical to update global reduction variables

### ISSUE THREAD

**Issue by wence-** [OP2/PyOP2#290]
If a parallel loop contains a reduction variable it the local contributions are incremented into it inside a critical block:

``` c

  #pragma omp parallel shared(boffset, nblocks, nelems, blkmap)
  {
    double arg0_0_l0[1][1]; // local reduction var
    for ( int i = 0; i < 1; i++ ) arg0_0_l0[0][i] = (double)0;
    #pragma omp for schedule(static)
    for ( int __b = boffset; __b < boffset + nblocks; __b++ )
    {
      for (int n = efirst; n < efirst+ nelem; n++ )
      {
...
    form_cell_integral_0_otherwise(arg0_0_l0[0], ...);
      }
    }
       // write back to global
      #pragma omp critical
      for ( int i = 0; i < 1; i++ ) arg0_0[i] += arg0_0_l0[0][i];
  }
```

This is horrible.  Each thread should write to a section of shared scratch space and then, after the iteration has completed, the master thread should do the final part of the reduction:

Git blame tells me this used to be the way things worked, but it changed in this commit:

```
commit dd0dfc179143d8cf40083c914e6960bcf4d3dc70
Author: Gheorghe-Teodor Bercea <gb308@doc.ic.ac.uk>
Date:   Wed Jun 12 18:36:36 2013 +0100

    Fix OpenMP globals. OpenMP and sequential are compatible with trunk.
```

We should think about undoing this.

**Comment by wence-** [comment]
No longer relevant after #507.
